src/gui/guiHyperText.cpp

```diff
@@ -0,0 +1,1069 @@
+		}
+
+		attrs[attr_name] = attr_val;
+	}
+
+	++cursor; // Last ">"
+
+	// Tag specific processing
+	StyleList style;
+
+	if (name == "global") {
+		if (end)
+			return 0;
+		globalTag(attrs);
+
+	} else if (name == "style") {
+		if (end)
+			closeTag(name);
+		else {
+			parseStyles(attrs, style);
+			openTag(name, attrs)->style = style;
+		}
+		endElement();
+	} else if (name == "img" || name == "item") {
+		if (end)
+			return 0;
+
+		// Required attributes
+		if (!attrs.count("name"))
+			return 0;
+
+		newTag(name, attrs);
+
+		if (name == "img")
+			enterElement(ELEMENT_IMAGE);
+		else
+			enterElement(ELEMENT_ITEM);
+
+		m_element->text = strtostrw(attrs["name"]);
+
+		if (attrs.count("float")) {
+			if (attrs["float"] == "left")
+				m_element->floating = FLOAT_LEFT;
+			if (attrs["float"] == "right")
+				m_element->floating = FLOAT_RIGHT;
+		}
+
+		if (attrs.count("rotate") && attrs["rotate"] == "yes")
+			m_element->rotation = IT_ROT_OTHER;
+
+		if (attrs.count("width")) {
+			int width = strtol(attrs["width"].c_str(), NULL, 10);
+			if (width > 0)
+				m_element->dim.Width = width;
+		}
+
+		if (attrs.count("height")) {
+			int height = strtol(attrs["height"].c_str(), NULL, 10);
+			if (height > 0)
+				m_element->dim.Height = height;
+		}
+		endElement();
+
+	} else if (name == "tag") {
+		// Required attributes
+		if (!attrs.count("name"))
+			return 0;
+
+		StyleList tagstyle;
+		parseStyles(attrs, tagstyle);
+
+		if (attrs["paragraph"] == "true")
+			m_paragraphtags[attrs["name"]] = tagstyle;
+		else
+			m_elementtags[attrs["name"]] = tagstyle;
+
+	} else if (name == "action") {
+		if (end)
+			closeTag(name);
+		else {
+			if (!attrs.count("name"))
+				return 0;
+			openTag(name, attrs)->style = m_elementtags["action"];
+		}
+
+	} else if (m_elementtags.count(name)) {
+		if (end)
+			closeTag(name);
+		else
+			openTag(name, attrs)->style = m_elementtags[name];
+		endElement();
+
+	} else if (m_paragraphtags.count(name)) {
+		if (end)
+			closeTag(name);
+		else
+			openTag(name, attrs)->style = m_paragraphtags[name];
+		endParagraph();
+
+	} else
+		return 0; // Unknown tag
+
+	// Update styles accordingly
+	m_style.clear();
+ 	for (auto tag = m_active_tags.crbegin(); tag != m_active_tags.crend(); ++tag)
+		for (auto const &prop : (*tag)->style)
+			m_style[prop.first] = prop.second;
+
+	return cursor;
+}
+
+// -----------------------------------------------------------------------------
+// Text Drawer
+
+TextDrawer::TextDrawer(
+	const wchar_t* text,
+	Client* client,
+	gui::IGUIEnvironment* environment,
+	ISimpleTextureSource *tsrc) :
+	m_text(text), m_client(client), m_environment(environment)
+{
+	// Size all elements
+	for (auto &p : m_text.m_paragraphs) {
+		for (auto &e : p.elements) {
+			if (e.type == ParsedText::ELEMENT_SEPARATOR ||
+					e.type == ParsedText::ELEMENT_TEXT) {
+				if (e.font) {
+					e.dim.Width = e.font->getDimension(e.text.c_str()).Width;
+					e.dim.Height = e.font->getDimension(L"Yy").Height;
```

Comment:
There are fontengine helper functions to determine the text dimensions.

Reply:
For some reasons `g_fontengine->getTextWidth(e.text.c_str());` and `g_fontengine->getTextHeight();` change the content positioning.

Reply:
The thing is we need to know the text dimension using its given font. I looked most appropriate to me.
If Truetype is in use, it looks like getDimention always gives the same height (not depending on string) so it could be simplified, but I'm not sure of the behavior if truetype not in use.